iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,183 @@
+    ATTR_ESCAPE_RE = re.compile('([' + MUST_ESCAPE_CHARS + '>"])')
+    STYLE_ESCAPE_RE = re.compile('([' + MUST_ESCAPE_CHARS + ']|' + CDATA_END + ')')
+
+    def __init__(self, fout, xml_declaration = True, assume_xhtml = True):
+        self.fout = fout
+        self.encoding = "utf-8"
+        self.xml_declaration = xml_declaration
+        self.assume_xhtml = assume_xhtml
+
+    def write(self, s):
+        self.fout.write(s.encode(self.encoding))
+
+    def prefix_sort(self, p):
+        return p if p is not None else '0'
+
+    def qname_for_node(self, node):
+        qname = etree.QName(node.tag)
+        if qname.namespace is None:
+            prefix = None
+        elif node.nsmap.get(node.prefix, None) == qname.namespace:
+            prefix = node.prefix
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
+        if prefix is None:
+            return qname.localname
+        return "%s:%s" % (prefix, qname.localname)
+
+    def qname_for_attr(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+        if qname.namespace == XML_NS:
+            prefix = 'xml'
+        else:
+            prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace and p is not None), key = self.prefix_sort)))
+        return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return self.ATTR_ESCAPE_RE.sub(lambda m: self.escape_str(m[0]), s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_str(self, c):
+        return self.ESCAPES.get(c, '&#x%02X;' % ord(c[0]))
+
+    def write_escape_text(self, s, escape_mode):
+        if s is None:
+            return
+        if escape_mode == EscapeMode.STYLE:
+            # HTML does not understand XML escapes inside a style element.
+            # Don't escape ">".  Escaping isn't required by XML, but is common
+            # practice, so we do it elsewhere.
+            # "&" and "<" may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            self.write(self.STYLE_ESCAPE_RE.sub(lambda m: self.escape_str(m[0]),s))
```

Comment:
Not a security risk, but it looks like CSS maybe has different escape syntax @paulwarren-wk ?

https://www.w3.org/International/questions/qa-escapes

I see there were some changes in the last 10 years, but latest chrome is giving me this.

```
>>> console.log(CSS.escape('<'));
\<
>>> console.log(CSS.escape('&'));
\&
>>> console.log(CSS.escape('\u0001'));
\1
```

Reply:
Ah - the comment there is out of date.  I've just committed an update.

The background here is that we're trying to serialize in a way that will yield a document that is both XHTML valid, and will render equivalently as XHTML and HTML, so we're trying to simultaneously comply with the different escaping rules applied by HTML and XHTML in `<style>` elements.

The issue that triggered this was that we were escaping `>` to `&gt;`, which is fine for XHTML, but not for HTML.

Escaping to `\<` and `\&` would still be XHTML invalid.  In earlier versions of this PR, we were escaping `<` and `&` using the `\NNNNNN` unicode syntax, on the basis that the only place those characters can currently legally appear are in strings, but we ultimately agreed to not do that, on the basis that a) there are proposals to add `&` to CSS syntax, and this escaping would break that and b) if HTML/XHTML compatibility was a goal of the original document author and these characters appeared in a string, they'd already be escaped.

Reply:
Okay, thanks for the info @paulwarren-wk . That new comment makes sense.